File: solutions/csharp/list-ops/ListOps.cs
using System;
using System.Collections.Generic;

public static class ListOps
{
    public static int Length<T>(List<T> input)
    {
        int count = 0;
        foreach (var _ in input)
        {
            count += 1;
        }
        return count;
    }

    public static List<T> Reverse<T>(List<T> input)
    {
        var final = new List<T>();
        foreach (var item in input)
        {
            final.Insert(0, item);
        }
        return final;
    }

    public static List<TOut> Map<TIn, TOut>(List<TIn> input, Func<TIn, TOut> map)
    {
        var final = new List<TOut>();
        foreach (var item in input)
        {
            final.Add(map(item));
        }
        return final;
    }

    public static List<T> Filter<T>(List<T> input, Func<T, bool> predicate)
    {
        var final = new List<T>();
        foreach (var item in input)
        {
            if (predicate(item)) { final.Add(item); }
        }
        return final;
    }

    public static TOut Foldl<TIn, TOut>(List<TIn> input, TOut start, Func<TOut, TIn, TOut> func)
    {
        var result = start;
        foreach (var item in input)
        {
            result = func(result, item);
        }
        return result;
    }

    public static TOut Foldr<TIn, TOut>(List<TIn> input, TOut start, Func<TIn, TOut, TOut> func)
    {
        var result = start;
        foreach (var item in Reverse(input))
        {
            result = func(item, result);
        }
        return result;
    }

    public static List<T> Concat<T>(List<List<T>> input)
    {
        var final = new List<T>();
        foreach (var list in input)
        {
            Append(final, list);
        }
        return final;
    }

    public static List<T> Append<T>(List<T> left, List<T> right)
    {
        var final = left;
        foreach (var item in right)
        {
            final.Add(item);
        }
        return final;
    }
}

File: solutions/swift/list-ops/Sources/ListOps/ListOps.swift
// append: given two lists, add all items in the second list to the end of the first list
func append<T>(_ first: [T], _ second: [T]) -> [T] {
  return foldLeft(second, accumulated: first, combine: { $0 + [$1] })
}

// concat: given a series of lists, combine all items in all lists into one flattened list
func concat<T>(_ lists: [T]...) -> [T] {
  return foldLeft(lists, accumulated: [], combine: append)
}

// filter: given a list and a predicate, return the list of all items for which predicate(item) is True
func filter<T>(_ list: [T], predicate: (T) -> Bool) -> [T] {
  return foldLeft(list, accumulated: [], combine: { predicate($1) ? $0 + [$1] : $0 })
}

// length: given a list, return the total number of items within it
func length<T>(_ list: [T]) -> Int {
  return foldLeft(list, accumulated: 0, combine: { (count, _) in count + 1 })
}

// map: given a list and a function, return the list of the results of applying function(item) on all items
func map<T>(_ list: [T], function: (T) -> T) -> [T] {
  return foldLeft(list, accumulated: [], combine: { $0 + [function($1)] })
}

// foldLeft: given a list, initial accumulator, and a function, fold (reduce) each item
//           into the accumulator from the left using function(accumulator, item)
func foldLeft<TIn, TOut>(
  _ list: [TIn], accumulated: TOut, combine: (TOut, TIn) -> TOut
) -> TOut {
  if list.isEmpty { return accumulated }
  return foldLeft(
    Array(list.dropFirst()), accumulated: combine(accumulated, list[0]), combine: combine)
}

// foldRight: given a list, initial accumulator, and a function, fold (reduce) each item
//            into the accumulator from the right using function(item, accumulator)
func foldRight<TIn, TOut>(
  _ list: [TIn], accumulated: TOut, combine: (TIn, TOut) -> TOut
) -> TOut {
  return foldLeft(reverse(list), accumulated: accumulated, combine: { combine($1, $0) })
}

// reverse: given a list, return a list with all the original items, but in reversed order
func reverse<T>(_ list: [T]) -> [T] {
  return foldLeft(list, accumulated: [], combine: { [$1] + $0 })
}
